### ISSUE TITLE
About Generating Protein Sequence Embeddings with Your Model

### ISSUE THREAD

**Issue by BlenderWang9487** [evolutionaryscale/esm#2]
Hi!

Thank you for your great work. I would like to ask if your model can be used solely for generating protein sequence embedding. For example, given a protein sequence, is there a function that produces its embedding for downstream tasks such as similarity search or property prediction with a simple linear head?

If so, do you have an example script that I can refer to? Or is there a best practice for generating such embedding?

Thank you!

**Comment by santiag0m** [comment]
Sure, you can get embeddings out of ESM3!

```python
from esm.models.esm3 import ESM3
from esm.sdk.api import ESMProtein, SamplingConfig
from esm.utils.constants.models import ESM3_OPEN_SMALL


client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device="cuda")

# Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
protein = ESMProtein(
    sequence=(
        "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
        "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
        "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
    )
)
protein_tensor = client.encode(protein)

output = client.forward_and_sample(
    protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
)
print(output.per_residue_embedding.shape)
```

If you have a PDB file you can also load the protein directly:

```python
protein = ESMProtein.from_pdb("./1utn.pdb")
```

**Comment by BlenderWang9487** [comment]
Got it! 👍

**Comment by fulacse** [comment]
> Sure, you can get embeddings out of ESM3!
> 
> ```python
> from esm.models.esm3 import ESM3
> from esm.sdk.api import ESMProtein, SamplingConfig
> from esm.utils.constants.models import ESM3_OPEN_SMALL
> 
> 
> client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device="cuda")
> 
> # Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
> protein = ESMProtein(
>     sequence=(
>         "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
>         "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
>         "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
>     )
> )
> protein_tensor = client.encode(protein)
> 
> output = client.forward_and_sample(
>     protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
> )
> print(output.per_residue_embedding.shape)
> ```
> 
> If you have a PDB file you can also load the protein directly:
> 
> ```python
> protein = ESMProtein.from_pdb("./1utn.pdb")
> ```

If I understand, output shape is [num_amino_acid, dim_embed]. I want process multi proteins at a time to get a tensor shaping [batch_size, num_amino_acid, dim_embed]. How to make a batch?

**Comment by ddofer** [comment]
I'd suggest adding this as an example or even in the readme, it's gonna be a recurring question.
(Ideally, running on a large set of sequences, and with the over trained small model?).
Thanks!

**Comment by vuhongai** [comment]
> Sure, you can get embeddings out of ESM3!
> 
> ```python
> from esm.models.esm3 import ESM3
> from esm.sdk.api import ESMProtein, SamplingConfig
> from esm.utils.constants.models import ESM3_OPEN_SMALL
> 
> 
> client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device="cuda")
> 
> # Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
> protein = ESMProtein(
>     sequence=(
>         "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
>         "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
>         "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
>     )
> )
> protein_tensor = client.encode(protein)
> 
> output = client.forward_and_sample(
>     protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
> )
> print(output.per_residue_embedding.shape)
> ```
> 
> If you have a PDB file you can also load the protein directly:
> 
> ```python
> protein = ESMProtein.from_pdb("./1utn.pdb")
> ```

In case of the input is sequence only, like in this example, should the _per_residue_embedding_ calculated from **forward_and_sample** and the embedding directly from the forward pass differ? If I understood well the code [here](https://github.com/evolutionaryscale/esm/blob/e9f060214d19a6420a69cd1187fdaf81978791fc/esm/models/esm3.py#L676-L680), it should not be different, right? I however see that they are very different. Maybe I missed something, can anyone please help me to understand why?  

Here's my example:

```
from esm.models.esm3 import ESM3
from esm.sdk.api import (
    ESMProtein,
    SamplingConfig
)

protein = ESMProtein(
    sequence="FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
)
protein_tensor = model.encode(protein)

with torch.no_grad():
    output1 = model(sequence_tokens=protein_tensor.sequence[None]).embeddings.detach().cpu().numpy()[0]
    output2 = model.forward_and_sample(
        protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
    ).per_residue_embedding.detach().cpu().numpy()
```

**Comment by zas97** [comment]
Is there a way to add padding to the sequence before generating the embeddings?

**Comment by santiag0m** [comment]
@zas97 check here: https://github.com/evolutionaryscale/esm/issues/28#issuecomment-2218430459

**Comment by santiag0m** [comment]
@vuhongai We use the model forward inside the `forward_and_sample` endpoint. It is likely that the difference you are seeing comes from how we initialize the defaults.

We are working on it (see https://github.com/evolutionaryscale/esm/issues/27)

**Comment by Lservi** [comment]
The whole protein embedding is the `print(output.protein_tensor)` ?

**Comment by Emalude** [comment]
> > Sure, you can get embeddings out of ESM3!
> > ```python
> > from esm.models.esm3 import ESM3
> > from esm.sdk.api import ESMProtein, SamplingConfig
> > from esm.utils.constants.models import ESM3_OPEN_SMALL
> > 
> > 
> > client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device="cuda")
> > 
> > # Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
> > protein = ESMProtein(
> >     sequence=(
> >         "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
> >         "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
> >         "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
> >     )
> > )
> > protein_tensor = client.encode(protein)
> > 
> > output = client.forward_and_sample(
> >     protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
> > )
> > print(output.per_residue_embedding.shape)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If you have a PDB file you can also load the protein directly:
> > ```python
> > protein = ESMProtein.from_pdb("./1utn.pdb")
> > ```
> 
> If I understand, output shape is [num_amino_acid, dim_embed]. I want process multi proteins at a time to get a tensor shaping [batch_size, num_amino_acid, dim_embed]. How to make a batch?

Any update on this? Right now, the only way I can do this is by calling multiple time the model on each ESMProtein object, which I guess is much slower than batch processing.

**Comment by thnhan** [comment]
> Sure, you can get embeddings out of ESM3!
> 
> from esm.models.esm3 import ESM3
> from esm.sdk.api import ESMProtein, SamplingConfig
> from esm.utils.constants.models import ESM3_OPEN_SMALL
> 
> 
> client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device="cuda")
> 
> # Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
> protein = ESMProtein(
>     sequence=(
>         "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
>         "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
>         "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
>     )
> )
> protein_tensor = client.encode(protein)
> 
> output = client.forward_and_sample(
>     protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
> )
> print(output.per_residue_embedding.shape)
> If you have a PDB file you can also load the protein directly:
> 
> protein = ESMProtein.from_pdb("./1utn.pdb")

Hi @santiag0m, I got the embeddings according to your guides as follows:

```
tensor([[[  48.7106,   71.4311,  -73.3834,  ...,   40.5793,   -1.3499,
            75.9258],
         [  22.7604,  161.1303, -101.0081,  ...,  -68.9090,   89.3172,
            20.6925],
         [ -12.6630, -165.3241,   98.5930,  ...,  -28.6245,  101.9472,
           -17.9971],
         ...,
         [ -23.1314,   29.0957,  -25.9369,  ...,  117.3580,   54.5923,
            40.4595],
         [ 125.5746,   76.5865,  218.1328,  ...,   99.1736,  127.3055,
           -13.1370],
         [   0.2512,  113.8707,  150.1742,  ...,   52.6727,   46.2013,
           106.2728]]])
```
But I am interested in why the values ​​of the embedding elements are larger than the embeddings generated by different modes like ProtT5, ESMC etc. I think those values ​​should be smaller, I mean the absolute value of those values ​​is in [0,1]. Could you guide me that I can use this kind of embedding for downstream tasks or is there any other way to generate embeddings using ESM3? Thank you!

**Comment by shijiale0609** [comment]
How to do structure embedding? Given a protein's multiple structures (ground state and multiple metastable states) and get the embedding for all these 3D structures.
